File: ui/js/contacts.js
/*
	contacts.js - manage all contact interactions  
	
	FUNCTIONS
		-
		-
*/
$(document).ready( function() {
	
	// DELETE A CONTACT
	$(".contact_delete").live("click", function() {
		var href 	= $(this).attr("href");
		var id		= $(this).attr("id");
		
		deleteparams = {
			"dream.in.verb" : "DELETE"
		};
		
		var deleteuri = apiuri(href,deleteparams);
		
		$.ajax({
			type: "POST",
			crossDomain:true, 
			url: deleteuri,
			success: function(results)
			{	
				$("#" + id).slideUp();
				closeModal();	
			}
		});
		
		return false;	
	});
	
	
	// ADD A NEW CONTACTS  
	$(".contact_post").live("submit", function() {
		
		var form	= $(this);
		var id 		= $(this).attr("id");
		var uri		= (api.root + api.contacts);
		
		clearstart(form);
		var data = get_contact_inputs(form);
		
		$.ajax({
			type: "POST",
			crossDomain:true,
			data: data,
			dataType: "json",
			mimeType: 'application/json',
			contentType: 'application/json',
			url: uri,
			success: function(results)
			{
				loadusercontacts();
				closeModal();	
			},
			error: function(results) 
			{
				console.log("failed");	
			}
		});	
		
		return false;		
	});
	
	// FORMAT PHONE NUMBERS
	$(".contact_phonenumber").live("keyup", function() {
		var number = $(this).val().replace(/(\d{3})(\d{3})(\d{4})/, "($1)-$2-$3");	
		$(this).val(number);
	});
	
	
	// EDIT EXISTING CONTACTS  
	$(".contact_put").live("submit", function() {
		var form	= $(this);
		var id 		= $(this).attr("id");
		var uri		= (api.root + api.contacts);
		var href 	= $(this).attr("href");
		
		// PREP THE FORM
		process(form);
		clearstart(form);
		var data = get_contact_inputs(form);
		startset();
					
		updateparams = {
			"dream.in.verb" : "PUT",
			"dream.out.format" : "json"
		};
		
		var updateuri = apiuri(href,updateparams);
		
		console.log(data);
		
		$.ajax({
			type: "POST",
			crossDomain:true,
			data: data,
			dataType: "json",
			mimeType: 'application/json',
			contentType: 'application/json',
			url: updateuri,
			success: function(results)
			{
				loadusercontacts();
				success(form);
			},
			error:function (xhr, ajaxOptions, thrownError){
				console.log(xhr.statusText);
				console.log(xhr.status);
				console.log(thrownError);
			}  
		});	
		
		return false;		
	});	
});

function get_contact_inputs(form)
{
	data = {
		contact : {
			firstname:  		form.find(".contact_firstname").val(),
			lastname:	  		form.find(".contact_lastname").val(),
			"uri.avatar":		form.find(".contact_picture img").attr("src"),
			age:  				form.find(".contact_age").val(),
			gender:  			form.find(".contact_gender").val(),
			"location":  		form.find(".contact_locationgeneral").val(),
			email:  			form.find(".contact_email").val(),
			phone:  			form.find(".contact_phonenumber").val(),
			fax:  				form.find(".contact_faxnumber").val(),
			address:  			form.find(".contact_address").val(),
			city:	  			form.find(".contact_city").val(),
			state:  			form.find(".contact_state").val(),
			zip:	  			form.find(".contact_zip").val(),
			"identity.twitter": 	form.find(".contact_twitter").val(),
			"identity.facebook": 	form.find(".contact_facebook").val(),
			"identity.linkedin": 	form.find(".contact_linkedin").val(),
			"identity.google": 		form.find(".contact_google").val(),
			company : {
				name:				form.find(".company_name").val(),
				phone:				form.find(".company_phonenumber").val(),
				address:			form.find(".company_address").val(),
				city:				form.find(".company_city").val(),
				state:				form.find(".company_state").val(),
				zip:				form.find(".company_zip").val(),
				industry:			form.find(".company_industry").val(),
				revenue:			form.find(".company_revenue").val(),
				competitors:		form.find(".company_competitors").val()
			}
		}	
	}	
	
	return JSON.stringify(data);
}

File: api/usivity.core/Services/Logic/Contacts.cs
using System.Linq;
using MindTouch.Dream;
using MindTouch.Xml;
using Usivity.Connections.Twitter;
using Usivity.Data;
using Usivity.Entities;
using Usivity.Entities.Types;

namespace Usivity.Core.Services.Logic {

    public class Contacts : IContacts {

        //--- Fields ---
        private readonly ICurrentContext _context;
        private readonly IUsivityDataCatalog _data;
        private readonly IOrganizations _organizations;

        //--- Constructors ---
        public Contacts(IUsivityDataCatalog data, ICurrentContext context, IOrganizations organizations) {
            _context = context;
            _data = data;
            _organizations = organizations;
        }

        //--- Methods ---
        public Contact GetContact(string id) {
            return _data.Contacts.Get(id, _organizations.CurrentOrganization);
        }

        public Contact GetNewContact(XDoc info) {
            var contact = new Contact();
            UpdateContactInformation(contact, info);
            contact.AddOrganization(_organizations.CurrentOrganization);
            return contact;
        }

        public void UpdateContactInformation(Contact contact, XDoc info) {
            contact.FirstName = info["firstname"].Contents ?? contact.FirstName;
            contact.LastName = info["lastname"].Contents ?? contact.LastName;
            var avatar = info["uri.avatar"].Contents;
            if(!string.IsNullOrEmpty(avatar)) {
                XUri uri;
                XUri.TryParse(avatar, out uri);
                if(uri != null) {
                    contact.Avatar = uri;
                }
            }
            contact.Age = info["age"].Contents ?? contact.Age;
            contact.Gender = info["gender"].Contents ?? contact.Gender;
            contact.Location = info["location"].Contents ?? contact.Location;
            contact.Phone = info["phone"].Contents ?? contact.Phone;
            contact.Fax = info["fax"].Contents ?? contact.Fax;
            contact.Address = info["address"].Contents ?? contact.Address;
            contact.City = info["city"].Contents ?? contact.City;
            contact.State = info["state"].Contents ?? contact.State;
            contact.Zip = info["zip"].Contents ?? contact.Zip;
            contact.CompanyName = info["company/name"].Contents ?? contact.CompanyName;
            contact.CompanyPhone = info["company/phone"].Contents ?? contact.CompanyPhone;
            contact.CompanyFax = info["company/fax"].Contents ?? contact.CompanyFax;
            contact.CompanyAddress = info["company/address"].Contents ?? contact.CompanyAddress;
            contact.CompanyCity = info["company/city"].Contents ?? contact.CompanyCity;
            contact.CompanyState = info["company/state"].Contents ?? contact.CompanyState;
            contact.CompanyZip = info["company/zip"].Contents ?? contact.CompanyZip;
            contact.CompanyIndustry = info["company/industry"].Contents ?? contact.CompanyIndustry;
            contact.CompanyRevenue = info["company/revenue"].Contents ?? contact.CompanyRevenue;
            contact.CompanyCompetitors = info["company/competitors"].Contents ?? contact.CompanyCompetitors;

            // Email
            var email = info["email"].Contents;
            if(!string.IsNullOrEmpty(email)) {
                var emailIdentity = new Identity {
                    Id = email
                };
                contact.SetIdentity(Source.Email, emailIdentity);
            }

            // Twitter
            var twitter = info["identity.twitter"].Contents;
            if(!string.IsNullOrEmpty(twitter)) {
                var twitterIdentity = TwitterPublicConnection.GetInstance().GetIdentity(twitter);
                contact.SetIdentity(Source.Twitter, twitterIdentity);
            }

            // Facebook
            var facebook = info["identity.facebook"].Contents;
            if(!string.IsNullOrEmpty(facebook)) {
                var facebookIdentity = new Identity {
                    Name = facebook
                };
                contact.SetIdentity(Source.Facebook, facebookIdentity);
            }

            // LinkedIn
            var linkedIn = info["identity.linkedin"].Contents;
            if(!string.IsNullOrEmpty(linkedIn)) {
                var linkedInIdentity = new Identity {
                    Name = linkedIn
                };
                contact.SetIdentity(Source.LinkedIn, linkedInIdentity);
            }

            // Google Plus
            var google = info["identity.google"].Contents;
            if(!string.IsNullOrEmpty(google)) {
                var googleIdentity = new Identity {
                    Name = google
                };
                contact.SetIdentity(Source.Google, googleIdentity);
            }
        }

        public void SaveContact(Contact contact) {
            _data.Contacts.Save(contact);
        }

        public XDoc GetContactsXml() {
            var contacts = _data.Contacts.Get(_organizations.CurrentOrganization);
            var doc = new XDoc("contacts")
                .Attr("count", contacts.Count())
                .Attr("href", _context.ApiUri.At("contacts"));
            foreach(var contact in contacts) {
                doc.Add(GetContactXml(contact));
            }
            doc.EndAll();
            return doc;
        }

        public XDoc GetContactXml(Contact contact, string relation = null) {
            return contact.ToDocument(relation)
                .Attr("href", _context.ApiUri.At("contacts", contact.Id));
        }

        public XDoc GetContactVerboseXml(Contact contact, string relation = null) {
            return contact.ToDocumentVerbose(relation)
                .Attr("href", _context.ApiUri.At("contacts", contact.Id));
        }

        public void RemoveContact(Contact contact) {
            contact.RemoveOrganization(_organizations.CurrentOrganization.Id);
            SaveContact(contact);
        } 
    }
}
